src/validate.js

```diff
@@ -0,0 +1,36 @@
+import { isObject, toErrorList, mergeObjects } from "./utils";
+
+
+function createErrorHandler(formData) {
+  const handler = {
+    __errors: [],
+    addError(message) {
+      this.__errors = [...this.__errors, message];
+    },
+  };
+  if (isObject(formData)) {
+    return Object.keys(formData).reduce((acc, key) => {
+      return {...acc, [key]: createErrorHandler(formData[key])};
+    }, handler);
+  }
+  return handler;
+}
+
+function unwrapErrorHandler(errorHandler) {
+  return Object.keys(errorHandler).reduce((acc, key) => {
+    if (key === "addError") {
+      return acc;
+    } else if (key === "__errors") {
+      return {...acc, [key]: errorHandler[key]};
```

Comment:
I had assumed the output of unwrapping ``{a: {b: {__errors: ["an error on a.b"]}}}`` should be ``{a: {b: ["an error on a.b"]}}``. But thinking about it now, maybe you're trying to permit someone to add an error to an "internal" node like a user as well as "leaf" nodes like password fields.